torch/distributed/rpc/functions.py

```diff
@@ -0,0 +1,46 @@
+import functools
+
+def async_execution(fn):
+    r"""
+    A decorator for a function indicating that the return value of the function
+    is guaranteed to be a ``torch.futures.Future`` object and this function can
+    run asynchronously on the RPC callee. This is useful when this function's
+    execution needs to pause and resume due to, e.g., containing
+    :meth:`~torch.distributed.rpc.rpc_async` or waiting for other signals.
```

Comment:
Should we add that the RPC on the callee side responds to the caller only when the returned future is completed? This would make it clear that we're not actually returning a Future over the wire.

Also, should we mention the order of decorators here? Basically this one needs to be the outer most decorator.